### ISSUE TITLE
Certain keys don't work for swedish (/non-english?) keyboard settings

### ISSUE THREAD

**Issue by AhmadGhaemi** [godbout/kindaVim.blahblah#168]
![Screen Recording 2023-03-31 at 14 46 04 mov](https://user-images.githubusercontent.com/93923832/230026498-ee5104bf-cee9-4732-8802-cc6239c6e41c.gif)

See the gif. I use cmd+j as a shortcut for normal mode in the gif. English keyboard settings work fine, but not the Swedish ones.  There is no key for '/' on a swedish keyboard; instead the combo shift+7 is used. But pressing shift+7 doesn't take me to search mode, neither does pressing the swedish equivalence of the english '/' -button (which is the '-' button for us). Have not checked if this counts for other keyboard languages as well but I would assume so.

**Comment by godbout** [comment]
> Have not checked if this counts for other keyboard languages as well but I would assume so.

i've tested with a French layout and there's the issue too. so definitely something i need to investigate. thanks for reporting!

**Comment by godbout** [comment]
things have been a bit slow lately coz i'm moving to a new place beginning of May. workstation will be back then and work will be back full time on kV then!

**Comment by godbout** [comment]
gonna start checking this now, for kV54. sorry for the late!

**Comment by AhmadGhaemi** [comment]
Cool! No worries

**Comment by godbout** [comment]
> There is no key for '/' on a swedish keyboard; instead the combo shift+7 is used. But pressing shift+7 doesn't take me to search mode, neither does pressing the swedish equivalence of the english '/' -button (which is the '-' button for us).

digging into this now. but i'm getting confused because it seems that each language would have to be handled independently? to generate a `/` on a French keyboard it's `shift + .`. what about the Swedish one? i don't understand your `shift + 7` AND `-`.

**Comment by godbout** [comment]
> > There is no key for '/' on a swedish keyboard; instead the combo shift+7 is used. But pressing shift+7 doesn't take me to search mode, neither does pressing the swedish equivalence of the english '/' -button (which is the '-' button for us).
> 
> digging into this now. but i'm getting confused because it seems that each language would have to be handled independently? to generate a `/` on a French keyboard it's `shift + .`. what about the Swedish one? i don't understand your `shift + 7` AND `-`.

of course i got it as soon as i posted.
humm. confusing to me. not sure how i'm gonna be able to integrate that properly. i could do exceptions for a few languages but that's not gonna be sustainable.

**Comment by AhmadGhaemi** [comment]
Yeah I meant that we have a `-`-key where the U.S. keyboard has its `/`-key.

Sorry, no idea how to implement this. I haven't worked with stuff like this so I have no idea how it works. Are there any adapters, like plugins that can be used? idk.

**Comment by godbout** [comment]
> Yeah I meant that we have a `-`-key where the U.S. keyboard has its `/`-key.

yeah got it after using my brain a little.

> Sorry, no idea how to implement this. I haven't worked with stuff like this so I have no idea how it works. Are there any adapters, like plugins that can be used? idk.

i'm using something that already map proper key codes depending on the keyboard layout (https://github.com/Clipy/Sauce). now the issue is that a same key combination may mean different things in different layouts. like for the Swedish keyboard, the `/` is `shift 7`. but on a French keyboard `shift 7` is gonna mean something else. and same thing may happen on different layouts. the mapper i'm using helps well for letters, like if you have a French software layout on a English hardware keyboard, pressing the `q` will do send a `a`, which is what we want. the issue now for me is translating all the things that are not letters it seems, according to the different language layouts. seems like a mega pain 😂️

thinking out loud here. not sure yet where this goes. but thinking and typing it helps my process.

**Comment by AhmadGhaemi** [comment]
Leave an issue on Sauce 😄 nah, let me know if you figure something out

**Comment by godbout** [comment]
> Leave an issue on Sauce 😄 nah, let me know if you figure something out

i don't think it's a Sauce issue. Sauce is not done for that. but yeah will def investigate and let you know.
btw except `/` and `?`, any other keys that don't work properly on your Swedish board?

**Comment by godbout** [comment]
so after deeper investigation, if i get this correctly, i need to revamp my whole key combination system. it was done the way it is currently because it seems the opposite issue was found early on. see: https://github.com/godbout/kindaVim.docs/issues/12

so i'm gonna have to test with different software layout but also different hardware combinations. gonna be a big one.

**Comment by godbout** [comment]
btw temporarily you could remap the keys through `Karabiner-Elements`, when kV is in `Normal Mode` (or more). kV sends macOS Distributed Notifications (https://github.com/godbout/kindaVim.docs#apis) when it changes mode. you can catch them and act accordingly. that's what's supposed to be used rather than kV building a native remapping tool. (in your case tho, it's more of a "bug" so i should take care of this natively.)

**Comment by AhmadGhaemi** [comment]
> > Leave an issue on Sauce 😄 nah, let me know if you figure something out
> 
> i don't think it's a Sauce issue. Sauce is not done for that. but yeah will def investigate and let you know. btw except `/` and `?`, any other keys that don't work properly on your Swedish board?

`$` too unfortunately. 

Are `.` and `q` supposed to work for repeating and macros respectively? they don't even when I have the US keyboard on.

**Comment by godbout** [comment]
> `$` too unfortunately.

👌️ will see what i can do.

> Are `.` and `q` supposed to work for repeating and macros respectively? they don't even when I have the US keyboard on.

macros are not implemented. the list of moves implemented is here: https://github.com/godbout/kindaVim.docs#current-vim-implementation

you can also see the ones that are currently _repeatable_. if you have move that are not repeated, you can request them here: https://github.com/godbout/kindaVim.docs/issues/138

**Comment by luisherranz** [comment]
I am seeing the same issue with my Spanish keyboard. For example, instead capturing the `]` key, it's capturing `alt` and `+`. I don't really understand why; the rest of the software I have installed on this computer correctly captures the final keys.

**Comment by godbout** [comment]
> I am seeing the same issue with my Spanish keyboard. For example, instead capturing the `]` key, it's capturing `alt` and `+`. I don't really understand why; the rest of the software I have installed on this computer correctly captures the final keys.

can you remap this for now through Karabiner-Elements or Hammerspoon? thing is yeah, it's a little hard because kV doesn't handle its own buffer like other apps do, as it's across the whole system. there's different _software layouts_ to handle as well as different _hardware layouts_. the hardware ones work, the software ones don't (yet), because 1) it's hard 2) there's tools like KE that are made especially for that. in the [latest kV video](https://www.youtube.com/watch?v=j78WEiFTZHM&t=2s) i mentioned the future of kV, and the next big thing will be custom key remapping. when this is done you should be able to do whatever you want directly from kV without using third party apps, but that's gonna take a while.

**Comment by godbout** [comment]
i'm bumping this issue up my todo list, tho.

**Comment by luisherranz** [comment]
Ok, thanks!

**Comment by godbout** [comment]
> Ok, thanks!

but thanks for adding your voice to that issue. it does indeed affect a few non english software keyboard layouts and because i've add only very few people reporting (all the other ones expect this issue were hardware keyboards and been fixed), this went down the priority list. but now that you're reminding me of this, it 1) makes me curious about how many people had the issue and never reported 2) forces me to face that the native kV custom remap is really important (even tho it's doable with third party apps). so, thanks again!

**Comment by godbout** [comment]
@AhmadGhaemi @luisherranz handling different keyboard layouts coming soon. target is within the next couple of weeks.
